File: boj2533_sns.py
import sys
sys.setrecursionlimit(10**9)
input = sys.stdin.readline


def dfs(i):
    visited[i] = 1
    dp[i] = [0, 1]
    for j in graph[i]:
        if not visited[j]:
            dfs(j)
            dp[i][0] += dp[j][1]
            dp[i][1] += min(dp[j])


n = int(input())
graph = [[] for _ in range(n + 1)]
for _ in range(n - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

dp = [[0, 0] for _ in range(n + 1)]
visited = [0] * (n + 1)
dfs(1)
print(min(dp[1]))

File: boj2533_sns.cpp
#include <iostream>
#include <vector>
using namespace std;

const int max_n = 1000001;
int n, dp[max_n][2];
bool visited[max_n];
vector<int> graph[max_n];

void dfs(int i) {
    visited[i] = true;
    dp[i][0] = 0;
    dp[i][1] = 1;

    for (int j : graph[i]) {
        if (!visited[j]) {
            dfs(j);
            dp[i][0] += dp[j][1];
            dp[i][1] += min(dp[j][0], dp[j][1]);
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n;
    for (int i = 1; i < n; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    dfs(1);
    cout << min(dp[1][0], dp[1][1]) << endl;

    return 0;
}
